# **Stars Classification**

This is a dataset consisting of several features of stars.

Some of them are:

- Absolute Temperature (in K)
- Relative Luminosity (L/Lo)
- Relative Radius (R/Ro)
- Absolute Magnitude (Mv)
- Star Color (white,Red,Blue,Yellow,yellow-orange etc)
- Spectral Class (O,B,A,F,G,K,,M)
- Star Type **(Red Dwarf, Brown Dwarf, White Dwarf, Main Sequence , SuperGiants, HyperGiants)**
- Lo = 3.828 x 10^26 Watts (Avg Luminosity of Sun)
- Ro = 6.9551 x 10^8 m (Avg Radius of Sun)

Program ini bertujuan untuk mengklasifikasikan Spektral Kelas Bintang K dengan 6 kelas lainnya.

In [ ]:
# Code berikut untuk menginstall library yang dibutuhkan
%pip install pandas
%pip install numpy
%pip install scikit-learn

In [ ]:
# Code berikut untuk mengimport library yang dibutuhkan, yaitu pandas dan numpy
import pandas as pd
import numpy as np

In [ ]:
# Code berikut untuk membaca dataset yang akan digunakan, yaitu dataset bintang yang diambil dari link berikut
star = pd.read_csv('https://github.com/YBIFoundation/Dataset/raw/main/Stars.csv')

In [ ]:
# Code berikut untuk menampilkan 5 data teratas dari dataset
star.head()

Temperature (K)  Luminosity (L/Lo)  Radius (R/Ro)  Absolute magnitude (Mv)  \
0             3068           0.002400         0.1700                    16.12   
1             3042           0.000500         0.1542                    16.60   
2             2600           0.000300         0.1020                    18.70   
3             2800           0.000200         0.1600                    16.65   
4             1939           0.000138         0.1030                    20.06   

   Star type Star category Star color Spectral Class  
0          0   Brown Dwarf        Red              M  
1          0   Brown Dwarf        Red              M  
2          0   Brown Dwarf        Red              M  
3          0   Brown Dwarf        Red              M  
4          0   Brown Dwarf        Red              M

In [ ]:
# Code berikut untuk menampilkan informasi dari dataset
star.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 240 entries, 0 to 239
Data columns (total 8 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Temperature (K)          240 non-null    int64  
 1   Luminosity (L/Lo)        240 non-null    float64
 2   Radius (R/Ro)            240 non-null    float64
 3   Absolute magnitude (Mv)  240 non-null    float64
 4   Star type                240 non-null    int64  
 5   Star category            240 non-null    object 
 6   Star color               240 non-null    object 
 7   Spectral Class           240 non-null    object 
dtypes: float64(3), int64(2), object(3)
memory usage: 15.1+ KB


In [ ]:
# Code berikut untuk menampilkan semua kolom yang ada pada dataset
star.columns

Index(['Temperature (K)', 'Luminosity (L/Lo)', 'Radius (R/Ro)',
       'Absolute magnitude (Mv)', 'Star type', 'Star category', 'Star color',
       'Spectral Class'],
      dtype='object')

In [ ]:
# Code berikut untuk menampilkan deskripsi dari dataset, yaitu berupa statistik deskriptif dari kolom numerik dataset
star.describe()

Temperature (K)  Luminosity (L/Lo)  Radius (R/Ro)  \
count       240.000000         240.000000     240.000000   
mean      10497.462500      107188.361635     237.157781   
std        9552.425037      179432.244940     517.155763   
min        1939.000000           0.000080       0.008400   
25%        3344.250000           0.000865       0.102750   
50%        5776.000000           0.070500       0.762500   
75%       15055.500000      198050.000000      42.750000   
max       40000.000000      849420.000000    1948.500000   

       Absolute magnitude (Mv)   Star type  
count               240.000000  240.000000  
mean                  4.382396    2.500000  
std                  10.532512    1.711394  
min                 -11.920000    0.000000  
25%                  -6.232500    1.000000  
50%                   8.313000    2.500000  
75%                  13.697500    4.000000  
max                  20.060000    5.000000

In [ ]:
# Code berikut untuk menghitung jumlah data pada kolom 'Spectral Class'
star[['Spectral Class']].value_counts()

Spectral Class
M                 111
B                  46
O                  40
A                  19
F                  17
K                   6
G                   1
dtype: int64

Spektral Kelas dipetakkan dengan angka 1 jika kelasnya adalah A, B, F, O, K, G dan 0 jika kelasnya adalah M

In [ ]:
# Code berikut untuk mengubah kolom 'Spectral Class' menjadi kode numerik
star.replace({'Spectral Class':{'M':0, 'A':1, 'B':1, 'F':1, 'O':1, 'K':1, 'G':1 }}, inplace=True)

In [ ]:
# Code berikut untuk menghitung jumlah data pada kolom 'Star type'
star[['Star type']].value_counts()

Star type
0            40
1            40
2            40
3            40
4            40
5            40
dtype: int64

Star color dipetakkan dengan angka 0 jika warnanya adalah Red, 1 jika warnanya adalah Yellow, 2 jika warnanya adalah White, dan 3 jika warnanya adalah Blue.

In [ ]:
# Code berikut untuk mengubah kolom 'Star color' menjadi kode numerik
star.replace({'Star color':{ 'Red':0, 'Yellow':1, 'White':2, 'White ': 2, 'Blue ':3, 'Blue':3 }}, inplace=True)

In [ ]:
# Code berikut untuk menghitung jumlah data pada kolom 'Star color'
star[['Star color']].value_counts()

Star color        
0                     112
3                      56
Blue-white             26
Blue White             10
yellow-white            8
2                       7
Blue white              3
white                   3
Yellowish White         3
Whitish                 2
yellowish               2
Orange                  2
White-Yellow            1
Pale yellow orange      1
Yellowish               1
Blue-White              1
Blue white              1
Orange-Red              1
dtype: int64

Variabel dependen atau yang akan diuji adalah kolom Spectral Class sedangkan variabel independen atau yang mempengaruhi variabel dependen adalah kolom Temperature, Luminosity, Radius, dan Absoute magnitude.

In [ ]:
# Code berikut memilih kolom yang akan digunakan sebagai variabel dependen (y) dan independen (X)
y = star['Spectral Class']
X = star[['Temperature (K)', 'Luminosity (L/Lo)', 'Radius (R/Ro)',
       'Absolute magnitude (Mv)']]

Data yang dilatih dan diuji dibagi menjadi 80:20 dengan random_statenya adalah 200.

In [ ]:
# Code berikut untuk membagi dataset menjadi data training dan data testing
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size = 0.8, random_state = 200)

In [ ]:
# Code berikut untuk membuat model machine learning
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()

In [ ]:
# Code berikut untuk melatih model machine learning
model.fit(X_train,y_train)

LogisticRegression()

In [ ]:
# Code berikut untuk membuat prediksi menggunakan model machine learning
y_pred = model.predict(X_test)

In [ ]:
# Code berikut untuk menampilkan hasil prediksi
y_pred

array([0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1,
       1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0,
       1, 1, 1, 1])

Hasil prediksi model menujukkan bahwa model memprediksi semua data sebagai 1 (kelas A, B, F, O, K, G) karena jumlah data kelas 1 lebih banyak dari kelas 0 (M).

In [ ]:
# import function
from sklearn.metrics import confusion_matrix, classification_report

In [ ]:
# Code berikut untuk menampilkan confusion matrix
confusion_matrix(y_test,y_pred)

array([[14,  4],
       [ 1, 29]])

Dari confusion matrix, kita bisa melihat bahwa model kita memiliki akurasi yang cukup baik dengan nilai 14 dan 4 untuk True Positive dan True Negative serta 1 dan 29 untuk False Positive dan False Negative

In [ ]:
# Code berikut untuk menampilkan classification report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.93      0.78      0.85        18
           1       0.88      0.97      0.92        30

    accuracy                           0.90        48
   macro avg       0.91      0.87      0.88        48
weighted avg       0.90      0.90      0.89        48



Hasil dari classification report menunjukkan bahwa model memiliki akurasi sebesar 0.90 dan memiliki f1-score sebesar 0.85 dan 0.92 untuk kelas 0 dan 1.

**Now predict for Start Class**